core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java

```diff
@@ -0,0 +1,381 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.concurrent.LockMode;
+import alluxio.resource.LockResource;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A simple inode lock list.
+ */
+@NotThreadSafe
+public class SimpleInodeLockList implements InodeLockList {
+  private static final Logger LOG = LoggerFactory.getLogger(SimpleInodeLockList.class);
+
+  private static final int INITIAL_CAPACITY = 8;
+  private static final Edge ROOT_EDGE = new Edge(-1, "");
+
+  private final InodeLockManager mInodeLockManager;
+
+  /**
+   * Entries for each lock in the lock list. The entries always alternate between InodeEntry and
+   * EdgeEntry.
+   */
+  private List<Entry> mEntries;
+  /**
+   * The index of the first write lock entry. All locks before this index are read locks, and all
+   * locks after and including this index are write locks. If all locks are read locks,
+   * mFirstWriteLockIndex == mEntries.size().
+   */
+  private int mFirstWriteLockIndex;
+
+  /**
+   * Creates a new empty lock list.
+   *
+   * @param inodeLockManager manager for inode locks
+   */
+  public SimpleInodeLockList(InodeLockManager inodeLockManager) {
+    mInodeLockManager = inodeLockManager;
+    mEntries = new ArrayList<>(INITIAL_CAPACITY);
+    mFirstWriteLockIndex = 0;
+  }
+
+  @Override
+  public void lockInode(Inode inode, LockMode mode) {
+    mode = nextLockMode(mode);
+    if (!mEntries.isEmpty()) {
+      Preconditions.checkState(!endsInInode());
+      String lastEdgeName = ((EdgeEntry) lastEntry()).getEdge().getName();
+      Preconditions.checkState(inode.getName().equals(lastEdgeName),
+          "Expected to lock inode %s but locked inode %s", lastEdgeName, inode.getName());
+    }
+    addEntry(createInodeEntry(inode, mode));
+  }
+
+  @Override
+  public void lockEdge(Inode lastInode, String childName, LockMode mode) {
+    mode = nextLockMode(mode);
+    long edgeParentId = lastInode.getId();
+    if (!mEntries.isEmpty()) {
+      Preconditions.checkState(endsInInode());
```

Comment:
can we add messages for these preconditions?

Reply:
good idea, added